src/disk_cache.rs

```diff
@@ -0,0 +1,1064 @@
+            Self::Direct(object_store) => object_store.head(location).await,
+        }
+    }
+
+    pub async fn get_opts(
+        &self,
+        location: &Path,
+        opts: CacheableGetOptions,
+    ) -> object_store::Result<GetResult> {
+        match self {
+            Self::Cached(inner) => inner.cached_get_opts(location, opts).await,
+            Self::Direct(object_store) => {
+                let get_opts = GetOptions {
+                    range: opts.range,
+                    ..Default::default()
+                };
+                object_store.get_opts(location, get_opts).await
+            }
+        }
+    }
+
+    pub fn list(&self, prefix: Option<&Path>) -> BoxStream<'_, object_store::Result<ObjectMeta>> {
+        match self {
+            Self::Cached(inner) => inner.object_store.list(prefix),
+            Self::Direct(object_store) => object_store.list(prefix),
+        }
+    }
+
+    pub fn buf_writer(&self, location: Path) -> BufWriter {
+        BufWriter::new(self.object_store(), location)
+    }
+
+    /// TODO: make this private
+    pub fn object_store(&self) -> Arc<dyn ObjectStore> {
+        match self {
+            Self::Cached(inner) => inner.object_store.clone(),
+            Self::Direct(object_store) => object_store.clone(),
+        }
+    }
+}
+
+impl From<Arc<dyn ObjectStore + 'static>> for CacheableObjectStore {
+    fn from(object_store: Arc<dyn ObjectStore + 'static>) -> Self {
+        Self::Direct(object_store)
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone)]
+pub(crate) struct CacheableObjectStoreInner {
+    object_store: Arc<dyn ObjectStore>,
+    part_size: usize, // expected to be aligned with mb or kb
+    storage: Arc<dyn LocalCacheStorage>,
+}
+
+impl CacheableObjectStoreInner {
+    pub fn new(
+        object_store: Arc<dyn ObjectStore>,
+        root_folder: std::path::PathBuf,
+        part_size: usize,
+    ) -> Self {
+        assert!(part_size % 1024 == 0);
+
+        let storage = Arc::new(DiskCacheStorage {
+            root_folder: root_folder.clone(),
+        });
+        Self {
+            object_store,
+            part_size,
+            storage,
+        }
+    }
+
+    pub async fn cached_head(&self, location: &Path) -> object_store::Result<ObjectMeta> {
+        let entry = self.storage.entry(location, self.part_size);
+        match entry.read_head().await {
+            Ok(Some((meta, _))) => Ok(meta),
+            _ => {
+                let result = self
+                    .object_store
+                    .get_opts(
+                        location,
+                        GetOptions {
+                            range: None,
+                            head: true,
+                            ..Default::default()
+                        },
+                    )
+                    .await?;
+                let meta = result.meta.clone();
+                self.save_result(result).await.ok();
```

Comment:
is there a way in rust to kick of this `save_result`, don't wait for it, and return? `save_result` should still continue to run to completion. I am new to rust, and newer to async in rust. aync only runs when awaited on is what I understand at a high level. Wonder if there is a way to make this happen.
There is some benefit in returning before local write completes.

Also, when disk cache is enabled, but we get error on disk writes, is the expectation to fail the writes or is it to skip the cache, provide an api to know that this is happening, and continue working?

I think continuing to work even when local disk writes fail would be preferable.

Reply:
it's a good point, i think we can spawn a tokio task over it, but spawning tasks may also prone to make the program OOM if the concurrency is not being throttled well.

i think this can be implemented as letting a standalone task to poll the "save task" from a channel, and maybe can be tried in another PR? this pr has already got a `XXL` label and i'm afraid of there has already backlogged many issues to be resolved :cry:

> Also, when disk cache is enabled, but we get error on disk writes, is the expectation to fail the writes or is it to skip the cache, provide an api to know that this is happening, and continue working?
>
> I think continuing to work even when local disk writes fail would be preferable.

yes, this pr tries to ignore every disk fail, and just pass back to the backend storage whenever any io error occurs. this is archived by the `.ok()` here to transform the result to an `Option` to tick the linter not complain about an `Result` not being processed.

i think i need match all the disk-failure related `Result` with some warning logs output.

Reply:
> maybe can be tried in another PR

+1